sundials@6.6.1 +sycl build fails: sunmemory_sycl.h:20:10: fatal error: 'CL/sycl.hpp' file not found

### Steps to reproduce the issue

`sundials@6.6.1 +sycl +examples-install cxxstd=17` fails to build using:
* `spack@develop` (f9c0a15ba058e63a6dbcc9674184438687313464 from `Tue Oct 31 12:28:16 2023 -0700`)
* OneAPI 2023.2.1 Toolkit
* Ubuntu 22.04

Spack Environment: [spack.yaml.txt](https://github.com/spack/spack/files/13221134/spack.yaml.txt)

<details><summary>Concretization</summary>

<pre>
 -   sundials@6.6.1%oneapi@2023.2.1+ARKODE+CVODE+CVODES+IDA+IDAS+KINSOL~cuda+examples+examples-install~f2003~fcmix+generic-math~ginkgo~hypre~int64~ipo~klu~kokkos~kokkos-kernels~lapack~magma~monitoring+mpi~openmp~petsc~profiling~pthread~raja~rocm+shared+static~superlu-dist~superlu-mt+sycl~trilinos build_system=cmake build_type=Release cstd=99 cxxstd=17 generator=make logging-level=0 logging-mpi=OFF precision=double arch=linux-ubuntu22.04-x86_64
[+]      ^cmake@3.27.7%oneapi@2023.2.1~doc+ncurses+ownlibs build_system=generic build_type=Release arch=linux-ubuntu22.04-x86_64
[+]          ^curl@8.4.0%oneapi@2023.2.1~gssapi~ldap~libidn2~librtmp~libssh~libssh2+nghttp2 build_system=autotools libs=shared,static tls=openssl arch=linux-ubuntu22.04-x86_64
[+]              ^nghttp2@1.57.0%oneapi@2023.2.1 build_system=autotools arch=linux-ubuntu22.04-x86_64
[+]              ^openssl@3.1.3%oneapi@2023.2.1~docs+shared build_system=generic certs=mozilla arch=linux-ubuntu22.04-x86_64
[+]                  ^ca-certificates-mozilla@2023-05-30%oneapi@2023.2.1 build_system=generic arch=linux-ubuntu22.04-x86_64
[+]                  ^perl@5.38.0%oneapi@2023.2.1+cpanm+opcode+open+shared+threads build_system=generic patches=714e4d1 arch=linux-ubuntu22.04-x86_64
[+]                      ^berkeley-db@18.1.40%oneapi@2023.2.1+cxx~docs+stl build_system=autotools patches=26090f4,b231fcc arch=linux-ubuntu22.04-x86_64
[+]                      ^bzip2@1.0.8%oneapi@2023.2.1~debug~pic+shared build_system=generic arch=linux-ubuntu22.04-x86_64
[+]                          ^diffutils@3.9%oneapi@2023.2.1 build_system=autotools arch=linux-ubuntu22.04-x86_64
[+]                              ^libiconv@1.17%oneapi@2023.2.1 build_system=autotools libs=shared,static arch=linux-ubuntu22.04-x86_64
[+]                      ^gdbm@1.23%oneapi@2023.2.1 build_system=autotools arch=linux-ubuntu22.04-x86_64
[+]                          ^readline@8.2%oneapi@2023.2.1 build_system=autotools patches=bbf97f1 arch=linux-ubuntu22.04-x86_64
[+]              ^pkgconf@1.9.5%oneapi@2023.2.1 build_system=autotools arch=linux-ubuntu22.04-x86_64
[+]          ^ncurses@6.4%oneapi@2023.2.1~symlinks+termlib abi=none build_system=autotools arch=linux-ubuntu22.04-x86_64
[+]          ^zlib-ng@2.1.4%oneapi@2023.2.1+compat+opt build_system=autotools arch=linux-ubuntu22.04-x86_64
[+]      ^gmake@4.4.1%oneapi@2023.2.1~guile build_system=generic arch=linux-ubuntu22.04-x86_64
[e]      ^mpich@4.1.2%oneapi@2023.2.1~argobots~cuda+fortran~hwloc+hydra+libxml2+pci~rocm+romio~slurm~vci~verbs~wrapperrpath build_system=autotools datatype-engine=auto device=ch4 netmod=ofi pmi=pmi arch=linux-ubuntu22.04-x86_64
</pre></details>

Error reproduced below in `ecpe4s/ubuntu22.04-runner-amd64-oneapi-2023.2.1-mpi-base:2023.10.19`
* Derived from Intel's official OneAPI container
```
$> spack -e . install
...
==> Installing sundials-6.6.1-jtom5qv7lyp6l35wnqyuev6gjfcrsju7 [9/9]
==> No binary for sundials-6.6.1-jtom5qv7lyp6l35wnqyuev6gjfcrsju7 found: installing from source
==> Using cached archive: /spack/var/spack/cache/_source-cache/archive/21/21f71e4aef95b18f954c8bbdc90b62877443950533d595c68051ab768b76984b.tar.gz
==> No patches needed for sundials
==> sundials: Executing phase: 'cmake'
==> sundials: Executing phase: 'build'
==> Error: ProcessError: Command exited with status 2:
    'make' '-j16'

7 errors found in build log:
...
     381    [  4%] Building C object src/sundials/CMakeFiles/sundials_generic_obj_static.dir/sundials_nvector.c.o
     382    cd /tmp/root/spack-stage/spack-stage-sundials-6.6.1-jtom5qv7lyp6l35wnqyuev6gjfcrsju7/spack-build-jtom5qv/src/sundials && /spack/lib/spack/env/oneapi/icx -DSUNDIALS_STATIC_DEF
            INE -I/tmp/root/spack-stage/spack-stage-sundials-6.6.1-jtom5qv7lyp6l35wnqyuev6gjfcrsju7/spack-src/include -I/tmp/root/spack-stage/spack-stage-sundials-6.6.1-jtom5qv7lyp6l35wn
            qyuev6gjfcrsju7/spack-build-jtom5qv/include -I/tmp/root/spack-stage/spack-stage-sundials-6.6.1-jtom5qv7lyp6l35wnqyuev6gjfcrsju7/spack-src/src/sundials -O3 -DNDEBUG -std=gnu99
             -fPIC -MD -MT src/sundials/CMakeFiles/sundials_generic_obj_static.dir/sundials_nvector.c.o -MF CMakeFiles/sundials_generic_obj_static.dir/sundials_nvector.c.o.d -o CMakeFile
            s/sundials_generic_obj_static.dir/sundials_nvector.c.o -c /tmp/root/spack-stage/spack-stage-sundials-6.6.1-jtom5qv7lyp6l35wnqyuev6gjfcrsju7/spack-src/src/sundials/sundials_nv
            ector.c
     383    In file included from /tmp/root/spack-stage/spack-stage-sundials-6.6.1-jtom5qv7lyp6l35wnqyuev6gjfcrsju7/spack-src/src/sunmemory/sycl/sundials_sycl_memory.cpp:19:
  >> 384    /tmp/root/spack-stage/spack-stage-sundials-6.6.1-jtom5qv7lyp6l35wnqyuev6gjfcrsju7/spack-src/include/sunmemory/sunmemory_sycl.h:20:10: fatal error: 'CL/sycl.hpp' file not foun
            d
     385    #include <CL/sycl.hpp>
     386             ^~~~~~~~~~~~~
     387    In file included from /tmp/root/spack-stage/spack-stage-sundials-6.6.1-jtom5qv7lyp6l35wnqyuev6gjfcrsju7/spack-src/src/sunmemory/sycl/sundials_sycl_memory.cpp:19:
  >> 388    /tmp/root/spack-stage/spack-stage-sundials-6.6.1-jtom5qv7lyp6l35wnqyuev6gjfcrsju7/spack-src/include/sunmemory/sunmemory_sycl.h:20:10: fatal error: 'CL/sycl.hpp' file not foun
            d
     389    #include <CL/sycl.hpp>
     390             ^~~~~~~~~~~~~
...
  >> 407    make[1]: *** [CMakeFiles/Makefile2:4940: src/sunmemory/sycl/CMakeFiles/sundials_sunmemsycl_obj_shared.dir/all] Error 2
...
  >> 423    make: *** [Makefile:149: all] Error 2
```


### Error message

<details><summary>Error message</summary>

<pre>
... see above
</pre></details>


### Information on your system

* **Spack:** 0.21.0.dev0 (f9c0a15ba058e63a6dbcc9674184438687313464)
* **Python:** 3.10.12
* **Platform:** linux-ubuntu22.04-cascadelake
* **Concretizer:** clingo

### Additional information

[spack-build-env.txt](https://github.com/spack/spack/files/13221138/spack-build-env.txt)
[spack-build-out.txt](https://github.com/spack/spack/files/13221139/spack-build-out.txt)

@balos1 @cswoodward @gardner48 @wspear @tldahlgren 

### General information

- [X] I have run `spack debug report` and reported the version of Spack/Python/Platform
- [X] I have run `spack maintainers <name-of-the-package>` and **@mentioned** any maintainers
- [X] I have uploaded the build log and environment files
- [X] I have searched the issues of this repo and believe this is not a duplicate

--- a/share/spack/gitlab/cloud_pipelines/stacks/e4s-oneapi/spack.yaml
+++ b/share/spack/gitlab/cloud_pipelines/stacks/e4s-oneapi/spack.yaml
@@ -200,7 +200,7 @@ spack:
   - kokkos +sycl +openmp cxxstd=17 +tests +examples
   - kokkos-kernels build_type=Release %oneapi ^kokkos +sycl +openmp cxxstd=17 +tests +examples
   - slate +sycl
-  - sundials +sycl cxxstd=17 +examples_install
+  - sundials +sycl cxxstd=17 +examples-install
   - tau +mpi +opencl +level_zero ~pdt +syscall   # tau: requires libdrm.so to be installed
   # --
   # - ginkgo +oneapi                    # InstallError: Ginkgo's oneAPI backend requires theDPC++ compiler as main CXX compiler.

--- a/share/spack/gitlab/cloud_pipelines/stacks/e4s-oneapi/spack.yaml
+++ b/share/spack/gitlab/cloud_pipelines/stacks/e4s-oneapi/spack.yaml
@@ -200,11 +200,11 @@ spack:
   - kokkos +sycl +openmp cxxstd=17 +tests +examples
   - kokkos-kernels build_type=Release %oneapi ^kokkos +sycl +openmp cxxstd=17 +tests +examples
   - slate +sycl
+  - sundials +sycl cxxstd=17 +examples_install
   - tau +mpi +opencl +level_zero ~pdt +syscall   # tau: requires libdrm.so to be installed
   # --
   # - ginkgo +oneapi                    # InstallError: Ginkgo's oneAPI backend requires theDPC++ compiler as main CXX compiler.
   # - hpctoolkit +level_zero            # dyninst@12.3.0%gcc: /usr/bin/ld: libiberty/./d-demangle.c:142: undefined reference to `_intel_fast_memcpy'; can't mix intel-tbb@%oneapi with dyninst%gcc
-  # - sundials +sycl cxxstd=17          # sundials: include/sunmemory/sunmemory_sycl.h:20:10: fatal error: 'CL/sycl.hpp' file not found
 
   - py-scipy
 

--- a/var/spack/repos/builtin/packages/sundials/package.py
+++ b/var/spack/repos/builtin/packages/sundials/package.py
@@ -292,6 +292,12 @@ class Sundials(CMakePackage, CudaPackage, ROCmPackage):
     # fix issues with exported PETSc target(s) in SUNDIALSConfig.cmake
     patch("sundials-v5.8.0.patch", when="@5.8.0")
 
+    def flag_handler(self, name, flags):
+        if name == "cxxflags":
+            if self.spec.satisfies("+sycl"):
+                flags.append("-fsycl")
+        return (flags, None, None)
+
     # ==========================================================================
     # SUNDIALS Settings
     # ==========================================================================

